### PR TITLE
[JENKINS-39450] LibraryStep

### PR DESCRIPTION
**Pull request by jglick**
Follow-up to #10 to add a `library` step.

@reviewbybees esp. @kohsuke

### CODE REVIEW

**Comment by kohsuke** [comment]
This is great!!

**Comment by kohsuke** [comment]
This works without library configuration, too, right? IIUC, you can write something like `library "github.com/kohsuke/my-pipeline-lib"` without touching Jenkins.

**Comment by jglick** [comment]
> This works without library configuration, too, right?

Untested, but yes it should.

**Comment by jglick** [comment]
Looks like the `workflow-cps` 2.14 update broke the `src/**/*.groovy` usage; need to make it return some stub objects, now raw `Class`es. (Which we would need anyway for various reaons.)

The other test failure looks like it is just due to downgrading the `git` dependency to a non-`SNAPSHOT`; should be easily fixable.

**Comment by abayer** [comment]
So after bumping back to `git` 2.6.1-SNAPSHOT and bumping `workflow-cps` to 2.22, the tests pass, but this still shows up in the `LibraryStepTest.classes` output:

```
Nov 02, 2016 10:14:04 AM org.jenkinsci.plugins.workflow.cps.CpsThreadGroup run
WARNING: program state save failed
java.io.IOException: Failed to persist /var/folders/pr/24nv8g910wg8vr4b4c33q34r0000gn/T/jenkinsTests.tmp/jenkins9048064252567545422test/jobs/p/builds/1/program.dat
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.saveProgram(CpsThreadGroup.java:444)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.saveProgram(CpsThreadGroup.java:412)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:357)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:78)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:236)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:224)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:63)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.NotSerializableException: pkg.App
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:860)
	at org.jboss.marshalling.river.BlockMarshaller.doWriteObject(BlockMarshaller.java:65)
	at org.jboss.marshalling.river.BlockMarshaller.writeObject(BlockMarshaller.java:56)
	at org.jboss.marshalling.MarshallerObjectOutputStream.writeObjectOverride(MarshallerObjectOutputStream.java:50)
	at org.jboss.marshalling.river.RiverObjectOutputStream.writeObjectOverride(RiverObjectOutputStream.java:179)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:343)
	at java.util.HashMap.writeObject(HashMap.java:1129)
	at sun.reflect.GeneratedMethodAccessor20.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.jboss.marshalling.reflect.SerializableClass.callWriteObject(SerializableClass.java:271)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:976)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:967)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.BlockMarshaller.doWriteObject(BlockMarshaller.java:65)
	at org.jboss.marshalling.river.BlockMarshaller.writeObject(BlockMarshaller.java:56)
	at org.jboss.marshalling.MarshallerObjectOutputStream.writeObjectOverride(MarshallerObjectOutputStream.java:50)
	at org.jboss.marshalling.river.RiverObjectOutputStream.writeObjectOverride(RiverObjectOutputStream.java:179)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:343)
	at java.util.TreeMap.writeObject(TreeMap.java:2274)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.jboss.marshalling.reflect.SerializableClass.callWriteObject(SerializableClass.java:271)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:976)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteFields(RiverMarshaller.java:1032)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteSerializableObject(RiverMarshaller.java:988)
	at org.jboss.marshalling.river.RiverMarshaller.doWriteObject(RiverMarshaller.java:854)
	at org.jboss.marshalling.AbstractObjectOutput.writeObject(AbstractObjectOutput.java:58)
	at org.jboss.marshalling.AbstractMarshaller.writeObject(AbstractMarshaller.java:111)
	at org.jenkinsci.plugins.workflow.support.pickles.serialization.RiverWriter.writeObject(RiverWriter.java:132)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.saveProgram(CpsThreadGroup.java:433)
	... 14 more
Caused by: an exception which occurred:
	in field locals
	in field parent
	in field caller
	in field e
	in field program
	in field threads
	in object org.jenkinsci.plugins.workflow.cps.CpsThreadGroup@78a4e690
```

**Comment by abayer** [comment]
...which, I suppose, is reasonable, since `pkg.App` isn't serializable?

**Comment by abayer** [comment]
Opened a JIRA for this - https://issues.jenkins-ci.org/browse/JENKINS-39450

**Comment by jglick** [comment]
> which, I suppose, is reasonable, since `pkg.App` isn't serializable?

Right, as I wrote on Sep 07, the current return value is untenable.

**Comment by abayer** [comment]
Gotcha.

**Comment by jglick** [comment]
> which, I suppose, is reasonable, since pkg.App isn't serializable?

Never mind, I totally missed your point here, and yes I corrected that in the test.

**Comment by jglick** [comment]
I believe this is functionally complete but I expect to flesh out the test coverage of corner cases before requesting general review.

File: src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryStepTest.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    @Test public void configRoundtrip() throws Exception {
+        StepConfigTester stepTester = new StepConfigTester(r);
+        SnippetizerTester snippetizerTester = new SnippetizerTester(r);
+        LibraryStep s = new LibraryStep("foo");
+        r.assertEqualDataBoundBeans(s, stepTester.configRoundTrip(s));
+        snippetizerTester.assertRoundTrip(s, "library 'foo'");
+        s.setRetriever(new SCMSourceRetriever(new GitSCMSource("id", "https://nowhere.net/", "", "*", "", true)));
+        r.assertEqualDataBoundBeans(s, stepTester.configRoundTrip(s));
+        // TODO should fix up common SCMSource implementations to use proper @DataBoundConstructor/@DataBoundSetter hygiene and add @Symbol
```
Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryStepTest.java]: Could you open a JIRA for that listing relevant `SCMSource` implementations you can think of? If so, I'll swoop in and clean them up accordingly.

**Comment by jglick** [comment]
> Could you open a JIRA for that listing relevant `SCMSource` implementations you can think of?

Well there is already [JENKINS-37227](https://issues.jenkins-ci.org/browse/JENKINS-37227). That is originally about `SCM` (thus helping `SCMRetriever`) rather than `SCMSource` (thus helping the preferred `SCMSourceRetriever`), but the tricky aspect is the same: investigating whether it is feasible to delete the `git` and `svn` custom steps in favor of the metastep, or whether we need to use alternate symbols for the `SCM` / `SCMSource` in those cases.

And it is not just a matter of adding a `@Symbol`, tricky as that may be for those two cases; you also need to rewrite the bloated `@DataBoundConstructor`s to have sensible defaults, and possibly offer simplified versions of the most crucial parameters (how many people _really_ need multiple `UserRemoteConfig`s / `ModuleLocation`s)? So I do not think this is a “swoop”.

**Comment by jglick** [comment]
Artifactory seems to be down.

**Comment by abayer** [comment]
@jglick It's flapping on and off. Good times.

**Comment by HRMPW** [comment]
:bee:
